text/0000-ok-wrapping.md

```diff
@@ -0,0 +1,548 @@
+- Feature Name: ok_wrapping
+- Start Date: 2017-06-04
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+[RFC 243]: https://github.com/rust-lang/rfcs/blob/master/text/0243-trait-based-exception-handling.md
+[RFC 1859]: https://github.com/rust-lang/rfcs/blob/master/text/1859-try-trait.md
+[RFC 1937]: https://github.com/rust-lang/rfcs/blob/master/text/1937-ques-in-main.md
+
+# Summary
+[summary]: #summary
+
+Add additional help for writing functions in an error handling mindset.
+Support a `?` on function definitions and `catch` blocks that automatically
+wraps the result in an `Ok`, `Some`, `Poll::Ready`, etc.  In particular,
+allow function authors to avoid the need for `Ok(())` at the end of
+functions returning `Result<(), E>`.
+
+(Note that this _ok wrapping_ only ever applies to the success case.  At no
+point would this allow a function to implicitly return `Err(e)`, `None`,
+`Poll::NotReady`, etc.)
+
+# Motivation
+[motivation]: #motivation
+
+The question mark operator ([RFC 243]) has made great strides in the
+ergonomics of error handling in Rust.  Having an explicit-but-short marker
+for fallible locations that propagates the error on failure allows function
+authors to operate in an error handling mindset, concentrating on the
+success path while acknowledging the failure path.
+
+There's still one place where this model breaks down, though: when
+generating the return value.
+
+Take this example from [RFC 1937]:
+
+```rust
+fn main() -> Result<(), Box<Error>> {
```

Comment:
Apparently it [does require C99](https://stackoverflow.com/a/3728153/431056), before that you get the "whatever was left around in the platform return register" behavior.